File: leetcode2/0061. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if k==0 or not head:
            return head
        p=head
        
        len_head=0
        while p:
            p=p.next
            len_head+=1
            
        k=k%len_head
        
        if k==0:
            return head
        
        node=ListNode(0)
        node.next=head
        p1=node
        for _ in range(len_head-k):
            p1=p1.next
        cur=p1.next
        p1.next=None
        
        p2=cur
        while p2.next:
            p2=p2.next
        p2.next=head
        
        return cur

File: archives/C#/0061. Rotate List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode RotateRight(ListNode head, int k) {
        if(k==0 || head==null){
            return head;
        }
        ListNode p=head;
        int cnt=0;
        while(p!=null){
            p=p.next;
            cnt++;
        }
        k=k%cnt;
        if(k==0){
            return head;
        }
        
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode p1=node;
        int keepLenth=cnt-k;
        while(keepLenth>0){
            p1=p1.next;
            keepLenth--;
        }
        ListNode newhead=p1.next;
        p1.next=null;
        
        ListNode pNewHead=newhead;
        while(pNewHead.next!=null){
            pNewHead=pNewHead.next;
        }
        pNewHead.next=head;
        return newhead;
    }
}
